Finer time unit for timeout.

jocelyn opened the pull request:
Added usecs support for main socket timeout (used by select_poll ...).
Added milliseconds support for recv and send timeout.

jocelyn committed:
Added configuration for Travis CI.

jocelyn committed:
Updated Travis CI config.

jocelyn committed:
Fixed Travis CI config to install EiffelStudio nightly.

jocelyn committed:
Fixed Travis CI config to install EiffelStudio nightly.

jocelyn committed:
Fixed Travis CI config to install EiffelStudio nightly.

jocelyn committed:
Fixed Travis CI config by using expected folder.

jocelyn committed:
Merge branch 'master' into develop

jocelyn committed:
Added usecs support for main socket timeout (used by select_poll ...).
Added milliseconds support for recv and send timeout.

jocelyn commented on .travis.yml:
```diff
@@ -0,0 +1,23 @@
+language: eiffel
```
Ignore this .travis.yml file.

jocelyn commented on Src/library/net/Clib/network.c:
```diff
@@ -1189,32 +1189,44 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+}
+
+void c_set_sock_recv_timeout_ms(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_milliseconds)
+	/* set socket SO_RCVTIMEO option to recv_timeout_milliseconds */
```
On Windows ... the C API accepts milliseconds, not usecs.

I made this patch a while ago .. but after thinking again at it
we should mimic routine for timeout (select_poll timeout) ..
to use 2 int values, one for the seconds, and another for usecs.
This way, if in the future, Windows accepts usecs, the Eiffel interface won't change, only the EiffelNet C code.
so it would be

void c_set_sock_recv_timeout_ms(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_seconds, EIF_INTEGER timeout_usec) {..}

kwaxer replied:
Some other socket API support nanoseconds - check recvmmsg, pselect, ppoll, etc. - they use nanoseconds for timeouts. Therefore I would go for Eiffel API that uses nanoseconds everywhere. Otherwise we'll have seconds, microseconds, milliseconds, nanoseconds - too many different time units. Seconds plus nanoseconds should be enough for everything.

Moreover, we can have a special Eiffel type that encapsulates those values and allows working on a higher level of abstraction.

jocelyn replied:
I agree after all.

jocelyn committed:
recv and send timeout setter now takes as timeout arguments values in seconds and microseconds.

jocelyn committed:
recv and send timeout setter now takes as timeout arguments values in seconds and nanoseconds.

jocelyn committed:
Improved C code to avoid unecessary conversion.
Note if timeout values are above zero, but less than smallest non zero timeout value
supported by the platform, then timeout is set to the smallest timeout value for the platform.

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -368,11 +439,11 @@ feature {NONE} -- Externals
 invariant
 
-	timeout_set: timeout > 0
+	timeout_set: timeout > 0 or (timeout_nano_seconds > 0 and timeout_nano_seconds < 1_000_000_000)
```
Shouldn't there be assertions _timeout >= 0_ and _timeout_nano_seconds >= 0_?
Also, the current invariant does not allow 0 timeout, is it expected?

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1176,8 +1176,28 @@ EIF_INTEGER c_set_sock_opt_linger(EIF_INTEGER fd, EIF_BOOLEAN flag, EIF_INTEGER
+	if (ns == 0) {
+		secs = 0;
+	} else {
+		secs = (EIF_INTEGER) (ns / RTI64C(1000000000)); /* Convert nanoseconds to seconds */
```
I gave a second thought on the logic and wonder if the calculation (here and at other places) could be simplified to
```
   us = (ns + 999) / 1_000
   ms = (ns + 999_999) / 1_000_000
   s = (ns + 999_999_999) / 1_000_000_000
```
This would be faster and also it would ensure the timeout is the specified value or a bit longer, but never shorter.

jocelyn committed:
Use a single nanoseconds value to set timeouts.
Removed all C routines, and replaced with new using only nanoseconds.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1176,8 +1176,12 @@ EIF_INTEGER c_set_sock_opt_linger(EIF_INTEGER fd, EIF_BOOLEAN flag, EIF_INTEGER
+/* timeout   											*/
+/********************************************************/
+
+EIF_NATURAL_64 c_get_sock_opt_timeo(EIF_INTEGER fd, EIF_INTEGER level, int optname)
```
If this is not exported to the Eiffel code, make it private, ie prefixing with `rt_private` or `static`.
Also I'm not a big fan of the name, so why not just `c_get_timeout_with_option_name`

jocelyn replied:
this was to follow other names such as `c_set_sock_opt_linger`

manuseiffel replied:
I actually do not care so much since the existing name are very confusing. At the very least it should be `timeout` and not `timeo`.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 	asize = sizeof(arg);
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, (char *) &arg, &asize));
```
remove the cast on optname, it is already an int.

manuseiffel replied:
No need for the cast to `char *` on `&arg` as it accepts a `void *`.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_RCVTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, (char *) &arg, &asize));
+	return RTU64C(1000000) * arg; /* 1 millisecond = 1000000 nanoseconds */
```
Can you add constants instead for all the variation of 10^x instead of having to put comments each time we use the value?

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 
-void c_set_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER recv_timeout_seconds)
-	/* set socket SO_RCVTIMEO option to recv_timeout_seconds */
+void c_set_sock_opt_timeo(EIF_INTEGER fd, EIF_INTEGER level, int optname, EIF_NATURAL_64 timeout_nano_seconds)
```
See comments about name above.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 #ifdef EIF_WINDOWS
-	int arg = (int) 1000 * recv_timeout_seconds; /* Timeout in milliseconds */
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_RCVTIMEO, (char *) &arg, sizeof(arg)));
+	int arg = (int)(timeout_nano_seconds / 1000000); /* Timeout in milliseconds: 1000000 nanoseconds =  1 millisecond */
```
Let's put an assert about the maximum value of timeout_nano_seconds to ensure that there will never be an overflow. Put the cooresponding require on the Eiffel side.

manuseiffel replied:
Why didn't you do the assert path since we already have the protection on the Eiffel side now?

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	int arg = (int) 1000 * recv_timeout_seconds; /* Timeout in milliseconds */
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_RCVTIMEO, (char *) &arg, sizeof(arg)));
+	int arg = (int)(timeout_nano_seconds / 1000000); /* Timeout in milliseconds: 1000000 nanoseconds =  1 millisecond */
+	if ((arg == 0) && (timeout_nano_seconds > 0)) {
```
No need to verify that timeout_nano_seconds is positive, it is already a NATURAL_64.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+			 */
+		arg = (int) 1;
+	}
+	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, (char *) &arg, sizeof(arg)));
```
remove the useless casts

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_RCVTIMEO, &tv, sizeof(struct timeval)));
+	tv.tv_sec = (long) (timeout_nano_seconds / RTU64C(1000000000)); 
+	tv.tv_usec = (long) (timeout_nano_seconds % RTU64C(1000)); /* 1 microsecond = 1000 nanoseconds */
+	if ((tv.tv_sec == 0 && tv.tv_usec == 0) && (timeout_nano_seconds > 0)) {
```
Remove check `timeout_nano_seconds > 0`.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1188,74 +1192,75 @@ EIF_INTEGER c_get_sock_recv_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+			 */
+		tv.tv_usec = (long) 1;
+	}
+	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) optname, &tv, sizeof(struct timeval)));
```
Remove useless casts

manuseiffel commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -107,15 +107,21 @@ EIF_INTEGER c_select_poll(EIF_INTEGER fd)
 
-	tmout.tv_sec = (unsigned long) timeout;
-	tmout.tv_usec = (long) 0;
+	tmout.tv_sec = (long) (timeout_nano_sec / RTU64C(1000000000));
```
We need to verify that the `timeout_nano_sec` is not going to overflow. On Windows `long` is 32-bit, whereas it is 64-bit on Linux.

jocelyn replied:
On the Eiffel side, I added a precondition to always stick to max 32bits.
So this should be safe.
However, I will also modify the C code to ensure no overflow occurs.

manuseiffel replied:
Just put an assertion

jocelyn replied:
maybe safer to have also on C side just in case someone is reusing this C code in the future directly.

manuseiffel commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -131,15 +137,21 @@ EIF_INTEGER c_select_poll_with_timeout(EIF_INTEGER fd,
 
-	tmout.tv_sec = (unsigned long) timeout;
-	tmout.tv_usec = (long) 0;
+	tmout.tv_sec = (long) (timeout_nano_sec / RTU64C(1000000000));
```
Check for overflow

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -141,7 +141,7 @@ feature -- Status report
 			retval: INTEGER
 		do
-			retval := c_select_poll_with_timeout (descriptor, True, timeout)
+			retval := c_select_poll_with_timeout (descriptor, True, {NATURAL_64} 1_000_000_000 * timeout.to_natural_64 + timeout_additional_nano_seconds.to_natural_64)
```
Use constants instead of hard-coded values

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,16 +163,33 @@ feature -- Status report
 	timeout: INTEGER
 			-- Duration of timeout in seconds
 
+	timeout_additional_nano_seconds: INTEGER
```
Let's use NATURAL_32 instead of INTEGER

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,16 +163,33 @@ feature -- Status report
+			-- Additional nanoseconds to `timeout`.
+			-- between 0 and < 1_000_000_000 nanoseconds = 1 second.
+
+	timeout_in_nano_seconds: NATURAL_64
```
I would have done the opposite. I would have made `timeout` an obsolete function returning the new attribute `timeout_ns` (shorter but still clear name instead of `timeout_in_nano_seconds`).

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -209,32 +234,63 @@ feature -- Status setting
 			timeout_set: timeout = n or timeout = default_timeout
 		end
 
+	set_fine_timeout (nb_nano_seconds: NATURAL_64)
```
I would have call it `set_timeout_ns` instead and mark `set_timeout` obsolete.

manuseiffel replied:
Add precondition to show that the argument is limited as the underlying C API only accepts 32-bit integers for seconds

manuseiffel reviewed (commented):
I think this is the right kind of change but let's hold on for 18.05 to commit them.

jocelyn committed:
Use a single timeout value using nanoseconds based on NATURAL_64.
Marked obsolete timeout features based on seconds.

jocelyn committed:
Check for overflow cases on long value.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, SO_SNDTIMEO, (char *) &arg, &asize));
-	return (EIF_INTEGER) (arg / 1000); /* Convert milliseconds to seconds */
+	eif_net_check(getsockopt((EIF_SOCKET_TYPE) fd, (int) level, optname, (char *) &arg, &asize));
+	return RTU64C(1000000) * arg; /* 1 millisecond = 1000000 nanoseconds */
```
What about having named constants like nanoseconds_in_millisecond, nanoseconds_in_second, etc.?

jocelyn replied:
done

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_SNDTIMEO, &tv, sizeof(struct timeval)));
+	tv.tv_sec = (long) (timeout_nano_seconds / RTU64C(1000000000)); 
+	tv.tv_usec = (long) (timeout_nano_seconds % RTU64C(1000)); /* 1 microsecond = 1000 nanoseconds */
```
Here we get the number of nanoseconds that do not fit a microsecond, not the number of microseconds.

jocelyn replied:
it will always be between 0 and 1000 , so fit usec long value.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 #ifdef EIF_WINDOWS
-	int arg = (int) 1000 * send_timeout_seconds; /* Timeout in milliseconds */
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_SNDTIMEO, (char *) &arg, sizeof(arg)));
+	int arg = (int)(timeout_nano_seconds / 1000000); /* Timeout in milliseconds: 1 millisecond = 1000000 nanoseconds */
```
Do we care if there resulting value does not fit into an int?

jocelyn replied:
done

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	tv.tv_usec = 0;
-
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_SNDTIMEO, &tv, sizeof(struct timeval)));
+	tv.tv_sec = (long) (timeout_nano_seconds / RTU64C(1000000000)); 
```
Do we care if the resulting value does not fit into a long?

jocelyn replied:
fixed

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -107,15 +107,27 @@ EIF_INTEGER c_select_poll(EIF_INTEGER fd)
+	} else {
+		tmout.tv_sec = (long) sec;
+	}
+	tmout.tv_usec = (long) (timeout_nano_sec % RTU64C(1000)); /* 1 microsecond = 1000 nanoseconds */
```
Same issue here - we get nanoseconds that don't fit into a microsecond.

jocelyn replied:
using modulo % ... the result of `timeout_nano_sec % RTU64C(1000)` will always be between 0 and 1000.
so I guess it will fit in tv_usec long value.

kwaxer replied:
Here is an example. `timeout_nano_sec` (BTW, would it be more consistent to use `timeout_ns` as in Eiffel code?) is equal to `2017` nanoseconds. This corresponds to 2 microseconds. But `2017 % 1000` gives `17`. This value will be assigned to `tv_usec` instead of `2`. I believe this is wrong.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1192,47 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-	eif_net_check (setsockopt((EIF_SOCKET_TYPE) fd, (int) level, (int) SO_SNDTIMEO, &tv, sizeof(struct timeval)));
+	tv.tv_sec = (long) (timeout_nano_seconds / RTU64C(1000000000)); 
+	tv.tv_usec = (long) (timeout_nano_seconds % RTU64C(1000)); /* 1 microsecond = 1000 nanoseconds */
+	if ((tv.tv_sec == 0 && tv.tv_usec == 0) && (timeout_nano_seconds > 0)) {
```
There are no parentheses in the similar test below, so I would remove them here to make the code consistent.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -131,15 +143,27 @@ EIF_INTEGER c_select_poll_with_timeout(EIF_INTEGER fd,
+	} else {
+		tmout.tv_sec = (long) sec;
+	}
+	tmout.tv_usec = (long) (timeout_nano_sec % RTU64C(1000)); /* 1 microsecond = 1000 nanoseconds */
```
Same here.

jocelyn replied:
same answer

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -141,7 +141,7 @@ feature -- Status report
 			retval: INTEGER
 		do
-			retval := c_select_poll_with_timeout (descriptor, True, timeout)
+			retval := c_select_poll_with_timeout (descriptor, True, timeout_ns)
```
What about removing the temporary local:
   `Result := c_select_poll_with_timeout (descriptor, True, timeout_ns) > 0`
?

jocelyn replied:
ok ... even if I find it easier when debugging to have such retval local variable...

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
+			"Use `timeout_ns` using nanoseconds [2018-01-15]"
+		do
+			Result := (timeout_ns // one_second_in_nanoseconds).to_integer_32
+			if Result = 0 then
```
A negative value is possible due to an overflow, so it might make sense to check that the result is non-negative.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
+		obsolete
+			"Use `recv_timeout_ns` using nanoseconds [2018-01-15]"
+		do
+			Result := (recv_timeout_ns \\ one_second_in_nanoseconds).to_integer_32
```
Should it be `//`?
Also, see a comment about possible overflow that results in a negative result.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
+		do
+			Result := (timeout_ns // one_second_in_nanoseconds).to_integer_32
+			if Result = 0 then
+					-- 0 may have different meaning, so return the lowest possible value.
```
What if the timeout was set to `0`?

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
+		obsolete
+			"Use `send_timeout_ns` using nanoseconds [2018-01-15]"
+		do
+			Result := (send_timeout_ns \\ one_second_in_nanoseconds).to_integer_32
```
Should it be `//`?
Also, see a comment about possible overflow.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
-			Result := c_get_sock_send_timeout (descriptor, level_sol_socket)
+			Result := c_get_sock_opt_timeo (descriptor, level_sol_socket, so_snd_timeo)
 		ensure
 			result_not_negative: Result >= 0
```
I guess this postcondition is now redundant.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -163,25 +163,57 @@ feature -- Status report
-			Result := c_get_sock_recv_timeout (descriptor, level_sol_socket)
+			Result := c_get_sock_opt_timeo (descriptor, level_sol_socket, so_rcv_timeo)
 		ensure
 			result_not_negative: Result >= 0
```
I guess this postcondition is now redundant.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
 
+	set_timeout_ns (a_timeout_nanoseconds: NATURAL_64)
+			-- Set timeout with to `a_timeout_nanoseconds` nanoseconds.
+			-- warning: the timeout granularity of the platform may not be nanoseconds, but micro or milliseconds.
```
**W**arning

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+			positive_timeout: a_timeout_nanoseconds /= 0
+			is_valid_timeout_ns: is_valid_timeout_ns (a_timeout_nanoseconds)
+		do
+			if a_timeout_nanoseconds = 0 then
```
This cannot happen because of the precondition.

jocelyn replied:
indeed, wrong precondition

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+			-- Set timeout with to `a_timeout_nanoseconds` nanoseconds.
+			-- warning: the timeout granularity of the platform may not be nanoseconds, but micro or milliseconds.
+		require
+			positive_timeout: a_timeout_nanoseconds /= 0
```
The tag name may need to be updated.

jocelyn replied:
on the other hand ... it is not wrong ... "positive" means > 0 
but ok "non_zero_timeout"

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+
 	set_recv_timeout (a_timeout_seconds: INTEGER)
 			-- Set the receive timeout in seconds on Current socket.
 			-- if `0' the related operations will never timeout.
```
Capitalization: **I**f...

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+		obsolete
+			"Use `set_recv_timeout_ns` using nanoseconds [2018-01-15]"
 		require
 			positive_timeout: a_timeout_seconds >= 0
```
It's non-negative.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+
+	set_recv_timeout_ns (a_timeout_nanoseconds: NATURAL_64)
+			-- Set the receive timeout with `a_timeout_nanoseconds` nanoseconds on Current socket.
+			-- if `0' the related operations will never timeout.
```
Capitalization at the beginning of a sentence.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+			-- if `0' the related operations will never timeout.
+			-- warning: the timeout granularity of the platform may not be nanoseconds, but micro or milliseconds.
+		require
+			positive_timeout: a_timeout_nanoseconds /= 0
```
The comment allows for `0` value.

jocelyn replied:
removed precondition

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
 	set_send_timeout (a_timeout_seconds: INTEGER)
-			-- Set the send timeout in milliseconds on Current socket.
+			-- Set the send timeout in seconds on Current socket.
 			-- if `0' the related operations will never timeout.
```
Capitalization.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+
+	set_send_timeout_ns (a_timeout_nanoseconds: NATURAL_64)
+			-- Set the send timeout with `a_timeout_nanoseconds` nanoseconds on Current socket.
+			-- if `0' the related operations will never timeout.
```
Capitalization.

jocelyn replied:
fixed

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -206,35 +238,73 @@ feature -- Status setting
+			-- if `0' the related operations will never timeout.
+			-- warning: the timeout granularity of the platform may not be nanoseconds, but micro or milliseconds.
+		require
+			positive_timeout: a_timeout_nanoseconds /= 0
```
Inconsistent with the comment.

jocelyn replied:
removed precondition

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -311,31 +381,49 @@ feature {NONE} -- Implementation
+feature -- Validation			
+
+	is_valid_timeout_ns (ns: NATURAL_64): BOOLEAN
+			--| note: internally (C API), the ns timeout is splitted into:
```
Capitalization.
Grammar: https://www.quora.com/Why-is-the-word-splitted-not-grammatically-correct

kwaxer replied:
What about having an official comment visible to users?

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -311,31 +381,49 @@ feature {NONE} -- Implementation
+			--|      and as NATURAL_64's max value is bigger     = 18_446_744_073_709_551_615
+			--|      features dealing with timeout in nanoseconds have related preconditions.
+		do
+			Result := 0 < ns and then ns <= max_timeout_ns_value
```
Is `0` timeout disabled? I understand that it may have a special meaning, but, possibly, C API still allows for `0` values. And, maybe, users do want to use this special behavior.

jocelyn replied:
In fact, in Current EiffelNet .. a timeout of 0 is prohibited , but I don't really see why.
I guess `0` means .. wait for ever, which may be wanted.

so I would remove that limitiation, and allow value 0 ... unless someone has a strong argument to keep it.

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -311,31 +381,49 @@ feature {NONE} -- Implementation
+	c_set_sock_opt_timeo (a_fd, level: INTEGER; a_timeo_opt: INTEGER; a_timeout_nanoseconds: NATURAL_64)
+		-- C routine to set socket option `$a_timeo_opt' with `a_timeout_nanoseconds`.
+		require
+			is_valid_timeout_ns: is_valid_timeout_ns (a_timeout_nanoseconds)
```
I guess, there are also restrictions on `level` and `a_timeo_opt`.

jocelyn replied:
I guess so as well.
This would be a part of another patch.
It is better to have small patch, rather than big one.

kwaxer reviewed (changes requested):
The most critical one seems to be the use of % operator.

jocelyn committed:
Fixed time computations (conversion or extraction).
Applied various suggestions.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1196,53 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+	int arg;
+	ms = timeout_ns / nanoseconds_in_millisecond; 
+	if (ms > INT_MAX) {
+		arg = (int) INT_MAX;
```
I guess the conversion `(int)` is useless here.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1196,53 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+			/* timeout values are above zero, but less than 1 millisecond
+			 * then set to 1 millisecond to avoid eventual special behavior for 0 timeouts).
+			 */
+		arg = (int) 1;
```
It could be just `1` without `(int)`.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1196,53 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+			/* timeout values are not zero, but less than 1 microsecond
+			 * then set to 1 microsecond to avoid eventual special behavior for 0 timeouts).
+			 */
+		tv.tv_usec = (long) 1;
```
The conversion `(long)` can be removed.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -94,7 +97,7 @@ EIF_INTEGER c_select_poll(EIF_INTEGER fd)
 	struct timeval timeout;
 
-	timeout.tv_sec = (unsigned long) 0;
+	timeout.tv_sec = (long) 0;
```
The conversion `(long)` can be removed.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1196,53 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
 
-
 EIF_INTEGER c_fcntl(EIF_INTEGER fd, EIF_INTEGER cmd, EIF_INTEGER arg)
 	/*x set possibly open fd socket options */
```
Several comments start from `/*x` instead of `/*`.

kwaxer commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,37 +1196,53 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
-void c_set_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level, EIF_INTEGER send_timeout_seconds)
-	/* set socket SO_SNDTIMEO option to send_timeout_seconds */
+void c_set_sock_opt_timeo(EIF_INTEGER fd, EIF_INTEGER level, int optname, EIF_NATURAL_64 timeout_ns)
+	/* set socket `optname` option to timeout_ns */
```
Other comments start from a capital letter, so it seems better to use the same style everywhere.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -107,15 +110,27 @@ EIF_INTEGER c_select_poll(EIF_INTEGER fd)
+			/* timeout values are above zero, but less than 1 microsecond
+			 * then set to 1 microsecond to avoid eventual special behavior for 0 timeouts).
+			 */
+		tmout.tv_usec = (long) 1;
```
The conversion `(long)` is not needed.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -107,15 +110,27 @@ EIF_INTEGER c_select_poll(EIF_INTEGER fd)
+	n64 = timeout_ns / nanoseconds_in_second;
+	if (n64 > LONG_MAX) {
+		/* Overflown, use max value instead */
+		tmout.tv_sec = (long) LONG_MAX;
```
The conversion `(long)` is not needed.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -131,15 +146,27 @@ EIF_INTEGER c_select_poll_with_timeout(EIF_INTEGER fd,
+	n64 = timeout_ns / nanoseconds_in_second;
+	if (n64 > LONG_MAX) {
+		/* Overflown, use max value instead */
+		tmout.tv_sec = (long) LONG_MAX;
```
The conversion `(long)` is not needed.

kwaxer commented on Src/library/net/Clib/syncpoll.c:
```diff
@@ -131,15 +146,27 @@ EIF_INTEGER c_select_poll_with_timeout(EIF_INTEGER fd,
+			/* timeout values are above zero, but less than 1 microsecond
+			 * then set to 1 microsecond to avoid eventual special behavior for 0 timeouts).
+			 */
+		tmout.tv_usec = (long) 1;
```
The conversion `(long)` is not needed.

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -194,47 +210,82 @@ feature -- Status setting
+		require
+			is_valid_timeout_ns: is_valid_timeout_ns (a_timeout_nanoseconds)
+		do
+			if a_timeout_nanoseconds = 0 then
```
I believe this case is ruled out by the precondition.

manuseiffel replied:
actually not since it is ok to have 0 per the precondition

kwaxer commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -309,33 +360,63 @@ feature {NONE} -- Implementation
+			if n64 > {INTEGER}.max_value then
+				Result := {INTEGER}.max_value
+			else
+				Result := (n64).to_integer_32
```
Unnecessary parentheses.

kwaxer reviewed (commented):
This version is almost ready.

jocelyn committed:
Removed a few unwanted cast.
Updated a few comments (style).

jocelyn commented:
I think most (if not all) comments are addressed.

manuseiffel commented on Src/library/net/Clib/network.c:
```diff
@@ -1225,39 +1196,55 @@ EIF_INTEGER c_get_sock_send_timeout(EIF_INTEGER fd, EIF_INTEGER level)
+	}
+	if (arg == 0 && timeout_ns != 0) {
+			/* timeout values are above zero, but less than 1 millisecond
+			 * then set to 1 millisecond to avoid eventual special behavior for 0 timeouts).
```
Do you have a reference for this as I do not understand this.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -116,76 +116,92 @@ feature -- Status report
 		do
-			reuse := c_get_sock_opt_int (descriptor, level_sol_socket, so_reuse_addr);
-			Result := reuse /= 0
+			Result := c_get_sock_opt_int (descriptor, level_sol_socket, so_reuse_addr) /= 0
```
I think the goal here was to use a local to see what the error code was.

jocelyn replied:
I agree as well, I changed to follow Alexander's suggestion.
Now it would even be better to store that value somewhere else, so that caller could check it too.
But this would be for another patch.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -116,76 +116,92 @@ feature -- Status report
 		do
-			retval := c_select_poll_with_timeout (descriptor, True, timeout)
-			Result := (retval > 0)
+			Result := c_select_poll_with_timeout (descriptor, True, timeout_ns) > 0
```
Same here, keep the local.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -116,76 +116,92 @@ feature -- Status report
 		do
-			retval := c_select_poll_with_timeout (descriptor, False, timeout)
-			Result := (retval > 0)
+			Result := c_select_poll_with_timeout (descriptor, False, timeout_ns) > 0
```
Same here, keep the local.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -116,76 +116,92 @@ feature -- Status report
 		do
-			retval := c_check_exception_with_timeout (descriptor, timeout)
-			Result := (retval > 0)
+			Result := c_check_exception_with_timeout (descriptor, timeout_ns) > 0
```
Same here, keep the local.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -116,76 +116,92 @@ feature -- Status report
+		do
+			Result := nanoseconds_to_seconds (timeout_ns)
+			if Result = 0 then
+					-- 0 may have different meaning, so return the lowest possible value.
```
I don't think we should be doing that but I don't know why you added this in the first place.

jocelyn replied:
I agree, but this is current behavior ... I search and find nothing that would explain why we should not accept 0 as timeout.
However here we should first check if timeout_ns is not zero.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -309,33 +362,63 @@ feature {NONE} -- Implementation
 		end
 
+	nanoseconds_to_seconds (ns: NATURAL_64): INTEGER
+		local
```
Put a comment and postcondition for the behavior.

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -309,33 +362,63 @@ feature {NONE} -- Implementation
-			-- C routine to set socket option `SO_RCVTIMEO' with `a_timeout_seconds' seconds.
-		external
-			"C"
+	one_second_in_nanoseconds: NATURAL_64 = 1_000_000_000
```
Comment

manuseiffel commented on Src/library/net/network/socket/network_socket.e:
```diff
@@ -309,33 +362,63 @@ feature {NONE} -- Implementation
+			--| Note: internally (C API), the ns timeout is splitted into:
+			--|  - a long value `tv_sec`: number of seconds
+			--|  - a long value `tv_usec`: number of microseconds (with struct timeval)
+			--|  or  long value `tv_nsec`: number of nanoseconds  (with struct timespec)
```
Where do we use `timespec` in the C API, I only see `timeval`? If we do not use `timespec` I would remove the comment.

jocelyn replied:
For now, we don't, but in the future we should use timespec, that is the reason we adopted nanosecond as timeout unit.
So I would keep this in comment.

jocelyn committed:
Added comment about zero timeout that could mean wait for ever.
Reverted to keep the local variable holding the return value (for debugging purpose).
Rename a C function to use full "timeout" instead of the C name "timeo".
Added missing comments and assertions.

jocelyn committed:
Added set_default_timeout and query is_default_timeout.
Apply changes to dependent classes.
Now `set_timeout (0)` really set the timeout to 0 and not to the default timeout.

jocelyn commented:
Clsoing as merged from subversion repo.
